[IV-21-22] Objetivo 7

Antobio17 opened the pull request:
Se ha trabajado en los puntos para la superación del objetivo 7.

- [x]  Creación de una entidad manejadora de configuración para la aplicación y archivos donde almacenar la misma (closes #40).
- [x]  Elección de sistema de logging para la aplicación y configuración del mismo (closes #41).
- [x]  Añadidos nuevos tests para comprobación de la nueva funcionalidad (closes #43).
- [x]  Añadida Github Action para la comprobación del dotenv (closes #45).

Antobio17 committed:
(#40) Entidad Config.py.
Se ha creado la entidad Config encargada de manejar los archivos de configuración del proyecto y cargarlos en objetos.

Antobio17 committed:
(#41) Registrando logs.
Se ha creado la entidad AppointmentScheduleManager encargada de registrar el log del funcionamiento de la reserva de citas. Se tendrán en cuenta el resultado de la operación y si se levantan excepciones.

Antobio17 committed:
(#43) Testeo de logging.
Se ha añadido un nuevo test para comprobar el correcto funcionamiento del sistema de logging implentado. También se comprobará el correcto funcionamiento de la entidad Config ya que está directamente relacionada.

Antobio17 committed:
(#41) Entidad logging.
Se ha creado la entidad logging encargada del registro de la aplicación.

Antobio17 committed:
Documentación del Sistema de logging...

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
(#40)(#41) Configuración.
Se crea el archivo de configuración encargado de almacenar la configuración pública del repositorio.
Se configura el logging para la aplicación.

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
(#45) Comprobación de dotenv.
Se trabaja en Github Action para comprobar que el dotenv está incluido en el archivo gitignore.

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
Eliminando .env para comprobar fallo de test.

Antobio17 committed:
(#45) Fix de Github Action.

Antobio17 committed:
(#45) Añadido .env a gitignore.

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube: Añadiendo .env a gitignore...

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
(#45) Se añade .env para testear...

Antobio17 committed:
(#45) Se elimina .env para testear...

Antobio17 committed:
(#40) Documentación...

Antobio17 committed:
:writing_hand:

Antobio17 committed:
:writing_hand:

Antobio17 committed:
Se añade clave/valor.

Antobio17 committed:
Fix iv.yaml

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
construction_worker_man:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
(#40) Se añade configuración por defecto.
Se añade configuració por defecto en método estático para la entidad Config.

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
Fix para logging en tests.
Se ha añadido la variable tests para que modificará los parámetros de configuración cuando se ejecuten los tests, por ejemplo para no sobre escribir el archivo de log.

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:test_tube:

Antobio17 committed:
:test_tube:

Antobio17 committed:
Se elimina fichero de configuración del repositorio remoto.

Antobio17 committed:
(#40)(#45) Refactorización.
Se usa config.yml para configuración de la aplicación.
Se elimina config.yml de repositorio y se modifica Github Action.

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
Se añade ETCD3 a dependencias.

Antobio17 committed:
Refactorización de entidad Config.py

Antobio17 committed:
Fix Naming

Antobio17 committed:
:construction_worker_man:

Antobio17 committed:
:writing_hand:

Antobio17 committed:
:writing_hand:

Antobio17 committed:
:writing_hand:

Antobio17 commented:
⛹ Revisores → @modejota @Slowmybrosh

Antobio17 committed:
Eliminación de entrada duplicada.

modejota commented on .github/workflows/config_checker.yml:
```diff
@@ -0,0 +1,44 @@
+    - uses: actions/checkout@v2
+      name: Checkout del repositorio 
+
+    - name: Comprobar que config.yml está añadido al gitignore
```
No entiendo demasiado el propósito de esta action, añadir el fichero de configuración al .gitignore es algo que haces una vez y muy probablemente no elimines, ¿para que comprobarlo cada vez?
Ojo, que me parece muy correcto para asegurarse, pero quizás reinventar la reuda.

Antobio17 replied:
Quiero evitar que el fichero de configuración este en el repositorio remoto.
Vi conveniente, además de comprobar si está o no commiteado el **YAML**, comprobar que la linea está añadida al _.gitignore_ y esta no se elimine.

JJ replied:
Realmente una vez que está añadido al push ya estás jodido... Sería mejor que lo tuvieras como un hook local.

modejota commented on docs/config.md:
```diff
@@ -0,0 +1,20 @@
+## Configuración local: config.yml
+Para almacenar la configuración se ha optado por usar un fichero **YAML** . Con los ficheros **YAML** tenemos la opción de construir estructuras de datos y definir tipos lo que nos dará una amplia posibilidad para almacenar la configuración de nuestra aplicación, como por ejemplo la configuración de logging.
+
+## Configuración remota
```
¿Has barajado alguna otra herramienta que potencialmente pudiera ofrecer ventajas frente a etcd3?

Antobio17 replied:
Si, he leido acerca de **Consul** pero para el uso que se le va a dar a la herramienta creo que **Etcd3** es mas adecuada por la simplicidad de uso una vez desplegada la aplicación.
Además en mi opinión la prioridad para obtener la configuración de la aplicación sería:
- 1. config.yml en local.
- 2. Configuración remota.
- 3. Configuración por defecto.

JJ replied:
Realmente, no es así. Tienes que poder usar las variables de entorno, porque es lo que se usa en los sistemas de CI/CD

modejota commented on my_hairdressing_app/appointment_schedule_manager.py:
```diff
@@ -0,0 +1,57 @@
+
+    def book_appointment(self, date: str, shift, appointment):
+        """
+        Fachada para manejar la función de reserva de citas de la clase AppointmentSchedule.
```
👍 por comentar las funciones.

Slowmybrosh commented:
Por lo que he leido parece estar todo bien, me ha gustado el check de que el .env está en el .gitignore y la documentación también estaba muy clara, tanto los requisitos que necesitas como la justificación de usar las herramientas. 

También le he echado un vistazo al código y parece estar todo bien y en orden, así que por mi parte no hay ninguna pega.

Antobio17 commented:
Muchas gracias por el trabajo a los revisores @modejota @Slowmybrosh 👍

Antobio17 commented:
@JJ listo para revisión

JJ commented on .github/workflows/config_checker.yml:
```diff
@@ -0,0 +1,44 @@
+      uses: actions/github-script@v3
+      with:
+        script: |
+          core.setFailed('❌ El fichero config.yml no está añadido en el .gitignore')
```
Le has pillado bien el tranquillo a las GitHub actions... Te será útil en entrevistas de trabajo.

JJ commented on azure-pipelines.yml:
```diff
@@ -41,6 +41,7 @@ steps:
 - script: |
     python -m pip install --upgrade pip
     pip install poetry
+    pip install pyyaml
```
Lo deberías instalar desde poetry

Antobio17 replied:
Si, de hecho lo modifique y arregle para instalarlo con mas dependencias de Poetry pero se me pasó hacerlo en los sistemas CI, lo arreglo.

JJ commented on docs/config.md:
```diff
@@ -0,0 +1,20 @@
+- Poder construir estructutas de datos mas complejas que las **clave/valor**.
+
+## Configuración local: config.yml
+Para almacenar la configuración se ha optado por usar un fichero **YAML** . Con los ficheros **YAML** tenemos la opción de construir estructuras de datos y definir tipos lo que nos dará una amplia posibilidad para almacenar la configuración de nuestra aplicación, como por ejemplo la configuración de logging.
```
Realmente aquí no hay que tomar muchas decisiones. Tienes que configurar vía variables de entorno sí o sí, y también mediente etcd3. Puedes usar un YAML de forma complementaria, pero eso no te va a servir en los sistemas de CI.

Antobio17 replied:
Mi intención era usar el **YML** como el que almacenara las variables de entorno, pero si no me sirve para los CI le doy una vuelta y paso a dotenv.

JJ commented on docs/logging.md:
```diff
@@ -0,0 +1,43 @@
+2. **Handlers**: en el que se le indicará uno de nuestros _formatters_ y podremos definir la salida del logging (consola, archivo, .etc).
+3. **Loggers**: en el que se indicará uno de los _handlers_ creados.
+
+El "inconveniente" del logging vanila de python es la necesidad de tener un fichero donde almacenar la configuración para cargarla posteriormente con una llamada a un método definido.
```
Uno de los requisitos debería ser que el sistema permita crear un objeto que se pueda inyectar en diferentes clases desde donde se vaya a usar. Incluso aunque uses el sistema por defecto, debes empaquetarlo (o instanciarlo) en un objeto.

Antobio17 replied:
Habia encapsulado la creación del log en mi clase Logging en la que devuelvo el logger creado con un método.

JJ commented on my_hairdressing_app/appointment_schedule_manager.py:
```diff
@@ -0,0 +1,57 @@
+        None
+        """
+        self.schedule = AppointmentSchedule()
+        self.logger = Logging(__name__, test).get_logger()
```
No estás usando la configuración, ni estás usando un objeto externo. Por *separation of concerns*, además, esta clase está encargada de crear el logger, lo que no es posible.

Antobio17 replied:
Vale creo que ya he entendido, yo lo que debo hacer es importar el logging ya configurado y usarlo directamente.

JJ reviewed (changes requested):
Una vuelta más y ya lo tienes.

Antobio17 committed:
(#41) Refactorización de Logging.
Se modifica la clase Logging para usar los metodos de log directamente desde el objeto importado sin necesidad de declarar una instancia de esta.
Fix en azure-pipelines.yml.

Antobio17 committed:
:construction:

Antobio17 committed:
(#40)(#45) Refactorización del manejo de configuración.
Se cambia el uso de config.yml a dotenv para declarar las valores de entorno de nuestro proyecto.
Se modifica la github action de checkeo de dotenv.

Antobio17 committed:
:construction:

Antobio17 committed:
Testeando variables de entorno en Circle CI...

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
Se añade módulo python-dotenv.

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
:construction:

Antobio17 committed:
Fixing...

Antobio17 committed:
:construction_worker__man:

Antobio17 committed:
:writing_hand:

Antobio17 committed:
:writing_hand:

Antobio17 commented:
Listo para revisión @JJ

JJ reviewed (approved):
:+1:
